### ISSUE TITLE
[Feature] IR-based numerical accuracy verification framework (IR Interpreter)

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#719]
### Summary

Add an IR interpreter that executes PyPTO IR programs op-by-op in Python (using PyTorch as the reference implementation), enabling numerical accuracy verification at compile time — **before any hardware codegen**. This framework serves two purposes:

1. **Operator correctness validation** — verify that each op's semantics description is numerically correct against PyTorch reference
2. **Pass correctness validation** — verify that IR transformations preserve numerical semantics by comparing execution results before and after a pass

This is a complex, multi-phase effort tracked below.

### Motivation / Use Case

Currently there is no systematic way to verify:
- Whether an operator's IR-level semantic description matches its intended mathematical behavior
- Whether a pass transformation preserves the numerical semantics of a program

Bugs in op definitions or passes are only caught late (during hardware execution or by manual inspection), leading to hard-to-debug silent correctness issues. An IR interpreter with PyTorch reference comparison would:
- Catch op semantic bugs **at definition time**, not at hardware execution time
- Catch pass correctness bugs **immediately after the pass runs**
- Enable automated regression testing of both ops and passes without hardware
- Provide a "ground truth" execution baseline for the entire compilation pipeline

### Design Overview

```
┌─────────────┐     ┌──────────────────┐     ┌─────────────────┐
│  PyPTO IR    │────▶│  IR Interpreter  │────▶│  Tensor results │
│  (program)   │     │  (op-by-op exec) │     │  (torch.Tensor) │
└─────────────┘     └──────────────────┘     └────────┬────────┘
                                                       │ compare
┌─────────────┐     ┌──────────────────┐     ┌────────▼────────┐
│  Same program│────▶│  PyTorch ref     │────▶│  Ref results    │
│  (or golden) │     │  execution       │     │  (torch.Tensor) │
└─────────────┘     └──────────────────┘     └─────────────────┘
```

**Core components:**
- **IR Interpreter**: walks the IR graph, dispatches each op to a registered Python handler that executes via PyTorch
- **Op Reference Registry**: maps each PyPTO op to its PyTorch reference implementation
- **Accuracy Comparator**: compares interpreter output vs reference with configurable tolerance (atol/rtol)
- **Pass Verifier**: wraps a pass, runs interpreter before & after, asserts numerical equivalence
- **Golden Snapshot**: optional mechanism to save/load reference outputs for regression testing

#### Phase 1: Core interpreter infrastructure
- [ ] Design `IRInterpreter` class that walks a `pl.program` and executes ops
- [ ] Implement value environment (SSA variable → `torch.Tensor` mapping)
- [ ] Support control flow: `for_range`, `if_then_else`, `grid` blocks
- [ ] Support function calls and nested scopes
- [ ] Handle tensor memory semantics (alloc, load, store, view)
- [ ] Add random input generation based on tensor type/shape annotations

#### Phase 2: Tensor-level op reference implementations
- [ ] Register PyTorch reference for core arithmetic ops (`add`, `sub`, `mul`, `div`, etc.)
- [ ] Register PyTorch reference for reduction ops (`sum`, `max`, `min`, `mean`, etc.)
- [ ] Register PyTorch reference for data movement ops (`transpose`, `reshape`, `slice`, `concat`, etc.)
- [ ] Register PyTorch reference for matmul / batch_matmul ops
- [ ] Register PyTorch reference for activation ops (`relu`, `gelu`, `silu`, `softmax`, etc.)
- [ ] Register PyTorch reference for normalization ops (`layer_norm`, `rms_norm`, etc.)
- [ ] Register PyTorch reference for comparison / selection ops (`where`, `clamp`, etc.)
- [ ] Add extensible registration mechanism for user-defined ops

#### Phase 3: Tile-level op reference implementations
- [ ] Register PyTorch reference for tile arithmetic ops
- [ ] Register PyTorch reference for tile data movement (load_tile, store_tile, etc.)
- [ ] Register PyTorch reference for tile reduction ops
- [ ] Register PyTorch reference for tile matmul (cube_matmul, etc.)
- [ ] Handle tile memory space semantics (L1, L0A, L0B, L0C, UB)
- [ ] Map tile-level indexing / slicing to equivalent torch operations

#### Phase 4: Accuracy comparison infrastructure
- [ ] Implement configurable comparator with `atol`, `rtol`, per-dtype defaults
- [ ] Support multiple comparison metrics: max abs error, mean abs error, cosine similarity
- [ ] Generate human-readable diff reports showing where mismatches occur
- [ ] Support per-op tolerance overrides (e.g., relaxed tolerance for softmax, matmul)
- [ ] Handle dtype-specific comparison (FP16 vs FP32 tolerance differences)

#### Phase 5: Pass correctness verification
- [ ] Implement `PassVerifier` wrapper: run interpreter before pass → run pass → run interpreter after → compare
- [ ] Support before/after numerical comparison mode
- [ ] Support golden reference snapshot mode (save expected outputs, compare on re-run)
- [ ] Integrate with `PassContext` to enable verification mode via configuration
- [ ] Add CLI/API option to enable pass verification in the pass pipeline

#### Phase 6: Testing and integration
- [ ] Add unit tests for the interpreter on each supported op
- [ ] Add pass verification tests for core passes (ConvertTensorToTileOps, FlattenTileNdTo2D, etc.)
- [ ] Add regression test infrastructure using golden snapshots
- [ ] Documentation: usage guide and op registration guide
- [ ] Integration with CI: run interpreter-based verification on selected test cases

### Proposed API / Behavior

```python
from pypto.verify import IRInterpreter, PassVerifier, AccuracyConfig

# --- 1. Basic interpretation ---
interp = IRInterpreter()
results = interp.run(program, inputs={"x": torch.randn(4, 128)})
# results: dict of output name → torch.Tensor

# --- 2. Op-level accuracy check ---
interp.verify_op("tensor.add", atol=1e-5, rtol=1e-5)

# --- 3. Pass correctness verification (before/after) ---
verifier = PassVerifier(
    interpreter=interp,
    config=AccuracyConfig(atol=1e-5, rtol=1e-5),
)
verifier.verify_pass(
    program,
    pass_fn=ConvertTensorToTileOps,
    inputs={"x": torch.randn(4, 128)},
)  # raises AccuracyError if mismatch

# --- 4. Golden snapshot mode ---
verifier.save_golden(program, inputs, path="golden/my_model.pt")
verifier.check_golden(program_after_pass, inputs, path="golden/my_model.pt")

# --- 5. Integrated in pass pipeline ---
with passes.PassContext(verification=True):
    pm.run_passes(program)  # each pass auto-verified
```

### Alternatives Considered

- **NumPy-based reference**: simpler but lacks GPU dtype semantics (FP16, BF16) and advanced ops that PyTorch provides natively
- **Hardware-based verification only**: requires actual NPU hardware, cannot run in CI without devices, much slower feedback loop
- **MLIR-style abstract interpretation**: more formal but significantly higher implementation complexity; PyTorch-based approach is more pragmatic and reuses existing ecosystem

### Additional Context

Related: #402 (IR print-parse roundtrip verification — structural correctness, complementary to this numerical correctness effort)
